addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpChannelState.java

```diff
@@ -0,0 +1,246 @@
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.TypeParser;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.http.internal.model.CommandRequest;
+import org.openhab.binding.http.internal.model.ErrorListener;
+import org.openhab.binding.http.internal.model.StateRequest;
+import org.openhab.binding.http.internal.model.Transform;
+import org.openhab.binding.http.internal.util.HttpUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+
+import static org.openhab.binding.http.internal.HttpBindingConstants.CHANNEL_TYPE_ID_IMAGE;
+import static org.openhab.binding.http.internal.HttpBindingConstants.DEFAULT_CONTENT_TYPE;
+
+/**
+ * This holds the state and handles commands for a channel connected to a HTTP Thing.
+ *
+ * @author Brian J. Tarricone - Initial contribution
+ */
+@NonNullByDefault
+public class HttpChannelState implements AutoCloseable {
+    private static final List<Class<? extends State>> STATE_TYPES = Arrays.asList(
+            OnOffType.class,
+            OpenClosedType.class,
+            UpDownType.class,
+            PointType.class,
+            HSBType.class,
+            DecimalType.class,
+            DateTimeType.class,
+            StringType.class
+    );
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final ChannelUID channelUID;
+    private final ChannelTypeUID channelTypeUID;
+    private final HttpClient httpClient;
+    private final int maxHttpResponseBodyLen;
+    private final Optional<StateRequest> stateRequest;
+    private final ScheduledExecutorService scheduler;
+    private final Optional<CommandRequest> commandRequest;
+    private final BiConsumer<ChannelUID, State> stateUpdatedListener;
+    private final ErrorListener errorListener;
+
+    private volatile boolean fetchingState = false;
+    private Optional<ScheduledFuture<?>> stateUpdater = Optional.empty();
+    private Optional<String> lastStateEtag = Optional.empty();
+
+    public HttpChannelState(final ChannelUID channelUID,
+                            final ChannelTypeUID channelTypeUID,
+                            final HttpClient httpClient,
+                            final int maxHttpResponseBodyLen,
+                            final Optional<StateRequest> stateRequest,
+                            final ScheduledExecutorService scheduler,
+                            final Optional<CommandRequest> commandRequest,
+                            final BiConsumer<ChannelUID, State> stateUpdatedListener,
+                            final ErrorListener errorListener)
+    {
```

Comment:
This constructor argument list is huge. Let's try to reduce it:
* A `HttpChannelState` is always tightly coupled to a handler, isn't it? Remove `httpClient`, `scheduler`, `errorListener`, `maxHttpResponseBodyLen` and replace it with the handler and grab all those values on demand from the handler.
* Provide the `Channel` instead of `channelUID` and `channelTypeUID` because you can deduce those two from a channel.

Reply:
Hmm, not sure I agree with passing in the handler.  I generally prefer to avoid circular dependencies between classes, as it makes things easier to reason about.

Reply:
I changed it to provide the `Channel` instead, at least.

Reply:
Also removed `maxHttpResponseBodyLen` and had `HttpChannelState` figure it out itself.

Reply:
If you want it clean you could extract necessary methods into an interface. But sticking to the separation of concerns pattern sometimes tangles objects together. In my opinion circular references are justified in that case.

You can as well have those two classes be merged, but you have split them on purpose. They still depend a lot on each other though, so why not go with the reference. @martinvw maybe you have an opinion on this?

Reply:
I guess the two classes could be merged, but IMO that would make things pretty messy.

Right now the `HttpChannelState` constructor takes 8 arguments.  I wouldn't consider that a lot (though agreed that it isn't a small number either).  Passing the handler in instead would reduce it to 4.  Alternatively, creating an interface for the 3 callback lambdas and passing an instance of that instead would give us 6.

I'm just not seeing how a small improvement on a (fairly arbitrary, IMO) metric like "number of arguments to constructor" justifies adding a circular class dependency or putting things behind an extra layer of abstraction.  To me, it just makes the code harder to read and understand.